docker/Dockerfile.deps

```diff
@@ -24,48 +25,39 @@ RUN yum remove -y devtoolset* && \
 RUN rm -f /opt/python/cp27-cp27m
 
 #zlib
-RUN LIBZ_VERSION=1.2.11 && \
-    cd /tmp && \
-    curl -L https://github.com/madler/zlib/archive/v${LIBZ_VERSION}.tar.gz | tar -xzf - && \
-    cd zlib-${LIBZ_VERSION} && \
-      CFLAGS="-fPIC" \
-      CXXFLAGS="-fPIC" \
+COPY third_party/zlib/ /tmp/zlib
+RUN cd /tmp/zlib && \
```

Comment:
`COPY` directives will invalidate the cache for **all** the subsequent layers, if the `<src>` part of the `COPY` changes. I believe, that every `git submodule update` will actually change the `<src>`, since the `.git` directory might change. That might impose the need of frequent rebuilding these changed Dockerfiles.

Maybe we could leverage multi-stage build to tackle this problem? Building every dependency in a separate stage, like:
```dockerfile
FROM ${FROM_IMAGE_NAME} as deps_builder
# ...

FROM deps_builder as dep_zlib
COPY third_part/zlib/ /tmp/zlib
RUN cd /tmp/zlib...

FROM deps_builder as dep_cmake
# etc... for all the dependencies

FROM ${FROM_IMAGE_NAME}
COPY --from=dep_zlib /usr/local/zlib /usr/local/zlib
COPY --from=dep_cmake ...
# etc...
```
This ensures, only the changed submodules will be rebuilt

Reply:
> that every git submodule update will actually change the <src>, since the .git directory might change.

The submodule itself doesn't have a `.git.` folder. Submodules history is kept in the main `.git/modules/PATH_TO_THE_MODULE/` folder. So Docker copy should work as it used to when we did the source downloads inside the docker.

`COPY --from=dep_zlib /usr/local/zlib /usr/local/zlib` is a terrible idea because of many side effects when you install given lib we may not be aware of (it installs headers, libs, pkg files, and other things). So there are multiple paths to track for every lib and it would be easy to miss things when we update dependencies.